### PR TITLE
jupyter ai with cborg

### PR DESCRIPTION
**Pull request by Tianhao-Gu**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+- JupyterLab 4
+
+## Install
+
+To install the extension, execute:
+
+```bash
+cd cdm-jupyter-ai-cborg
+pip install -e "."
+```
```
Comment (MrCreosote) [inline README.md:22]: Is there some context re where this install needs to happen?

File: jupyter_ai_cborg/_version.py
Comment (MrCreosote) [inline jupyter_ai_cborg/_version.py:1]: The notes say this file shouldn't be tracked in git

File: jupyter_ai_cborg/cborg.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            openai_api_base=cborg_api_base,
+            **kwargs,
+        )
+
+    @classmethod
+    def is_api_key_exc(cls, e: Exception):
+        """
+        Determine if the exception is an CBorg API key error.
+        """
+        import openai
```
Comment (MrCreosote) [inline jupyter_ai_cborg/cborg.py:60]: Why import here vs the top of the flie?

**Review by MrCreosote — COMMENTED** [review]
I don't really understand enough about how this works to review it effectively, but what I can glean from the code seems reasonable other than a few comments

File: jupyter_ai_cborg/_version.py
Comment (Tianhao-Gu) [reply jupyter_ai_cborg/_version.py:1]: 👍 removed.

File: jupyter_ai_cborg/cborg.py (hunk 2 above)
Comment (Tianhao-Gu) [reply jupyter_ai_cborg/cborg.py:60]: I reviewed the Jupyter-AI source code and found they were implementing it this way, so I just adapted it directly from their repository.

https://github.com/jupyterlab/jupyter-ai/blob/main/packages/jupyter-ai-magics/jupyter_ai_magics/partner_providers/openai.py#L21-L31

File: README.md (hunk 1 above)
Comment (Tianhao-Gu) [reply README.md:22]: hmmm this command is just for local installation. In our juptyerhub, we will pip install from git. Is that what you are trying to ask?

Comment (MrCreosote) [reply README.md:22]: For example, if I start up a new ubuntu 24 vm, will this do anything useful? What if I'm in a pipenv / poetry / *conda virtual env?

Comment (Tianhao-Gu) [reply README.md:22]: I don't think we need to go into those details and list all installation command for different virtual env setup.

Comment (MrCreosote) [reply README.md:22]: I think the instructions need to explain under what circumstances they work at minimum

File: jupyter_ai_cborg/_version.py
Comment (MrCreosote) [reply jupyter_ai_cborg/_version.py:1]: :+1:

File: jupyter_ai_cborg/cborg.py (hunk 2 above)
Comment (MrCreosote) [reply jupyter_ai_cborg/cborg.py:60]: :+1:

File: README.md (hunk 1 above)
Comment (Tianhao-Gu) [reply README.md:22]: I don't get what you are trying to ask here. The install instructions work if you have everything listed under `Requirements`.

Comment (MrCreosote) [reply README.md:22]: Do they work with pipenv / poetry / conda?

Comment (Tianhao-Gu) [reply README.md:22]: yea. I don't see why not.

Comment (MrCreosote) [reply README.md:22]: ok

**Review by MrCreosote — APPROVED** [review]

**Pull request merged**